language/move-prover/src/boogie_helpers.rs

```diff
@@ -0,0 +1,203 @@
+            boogie_struct_type_value(env, *module_id, *struct_id, args)
+        }
+        // TODO: function and tuple types?
+        Type::Tuple(_args) => "Tuple_type_value()".to_string(),
+        Type::Fun(_args, _result) => "Function_type_value()".to_string(),
+        _ => panic!("unexpected transient type"),
+    }
+}
+
+/// Create boogie type value for a struct with given type actuals.
+pub fn boogie_struct_type_value(
+    env: &GlobalEnv,
+    module_id: ModuleId,
+    struct_id: StructId,
+    args: &[Type],
+) -> String {
+    let module_env = env.get_module(module_id);
+    let struct_env = module_env.get_struct(struct_id);
+    format!(
+        "{}_type_value({})",
+        boogie_struct_name(&struct_env),
+        boogie_type_values(env, args)
+    )
+}
+
+/// Create boogie type value list, separated by comma.
+pub fn boogie_type_values(env: &GlobalEnv, args: &[Type]) -> String {
+    args.iter()
+        .map(|arg| boogie_type_value(env, arg))
+        .join(", ")
+}
+
+/// Return boogie type for a local with given signature token.
+pub fn boogie_local_type(ty: &Type) -> String {
+    if ty.is_reference() {
+        "Reference".to_string()
+    } else {
+        "Value".to_string()
+    }
+}
+
+/// Create boogie type check boolean expression.
+pub fn boogie_type_check_expr(env: &GlobalEnv, name: &str, ty: &Type) -> String {
+    let mut conds = vec![];
+    //let mut ret = String::new();
+    match ty {
+        Type::Primitive(p) => match p {
+            PrimitiveType::U8 => conds.push(format!("$IsValidU8({})", name)),
+            PrimitiveType::U64 => conds.push(format!("$IsValidU64({})", name)),
+            PrimitiveType::U128 => conds.push(format!("$IsValidU128({})", name)),
+            PrimitiveType::Num => conds.push(format!("$IsValidNum({})", name)),
+            PrimitiveType::Bool => conds.push(format!("is#Boolean({})", name)),
+            PrimitiveType::Address => conds.push(format!("is#Address({})", name)),
+            PrimitiveType::ByteArray => conds.push(format!("is#ByteArray({})", name)),
+            PrimitiveType::Range => conds.push(format!("$IsValidRange({})", name)),
+        },
+        Type::Vector(_) => conds.push(format!("$Vector_is_well_formed({})", name)),
+        Type::Struct(module_idx, struct_idx, _) => {
+            let struct_env = env.get_module(*module_idx).into_get_struct(*struct_idx);
+            conds.push(format!(
+                "${}_is_well_formed({})",
+                boogie_struct_name(&struct_env),
+                name
+            ))
+        }
+        Type::Reference(_is_mut, rtype) => {
+            conds.push(boogie_type_check_expr(
+                env,
+                &format!("$Dereference($m, {})", name),
+                rtype,
+            ));
+            conds.push(format!(
+                "$IsValidReferenceParameter($m, $local_counter, {})",
+                name
+            ));
+        }
+        // TODO: tuple and functions?
```

Comment:
Should we panic in these cases to avoid surprises?

Reply:
We actually report an error on an outer level.